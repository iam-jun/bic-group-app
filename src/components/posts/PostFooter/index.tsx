import React, { FC, useCallback } from 'react';

import { Omit } from 'react-native';
import { useRootNavigation } from '~/hooks/navigation';
import homeStack from '~/router/navigator/MainStack/stacks/homeStack/stack';
import ContentFooter, { ContentFooterProps } from '~/components/ContentView/ContentFooter';
import usePostsStore from '~/store/entities/posts';
import postsSelector from '~/store/entities/posts/selectors';
import { formatLargeNumber } from '~/utils/formatter';
import { useBaseHook } from '~/hooks';
import { ContentFooterLite, ContentFooterLiteProps } from '~/components/ContentView';
import { getTotalReactions } from '~/helpers/post';
import { PostType } from '~/interfaces/IPost';
import { trackEvent } from '~/services/tracking';
import { TrackingEventContentReadProperties } from '~/services/tracking/Interface';
import { TrackingEventContentReadAction, TrackingEvent } from '~/services/tracking/constants';

export interface PostFooterProps extends Partial<ContentFooterProps>,
 Partial<Omit<Omit<ContentFooterLiteProps, 'reactionsCount'>, 'onPressComment'>> {
  postId: string;
  isLite?: boolean;
}

const PostFooter: FC<PostFooterProps> = ({
  postId,
  isLite,
  onPressComment,
  ...props
}) => {
  const { rootNavigation } = useRootNavigation();
  const { t } = useBaseHook();

  const reactionsCount = usePostsStore(useCallback(postsSelector.getReactionCounts(postId), [postId]));
  const commentsCount = usePostsStore(useCallback(postsSelector.getCommentsCount(postId), [postId]));
  const setting = usePostsStore(useCallback(postsSelector.getSetting(postId), [postId]));
  const totalUsersSeen = usePostsStore(useCallback(postsSelector.getTotalUsersSeen(postId), [postId]));
  const {
    canComment, canReact,
  } = setting || {};

  const commentCount = formatLargeNumber(commentsCount);
  const labelButtonComment = `${commentCount ? `${commentCount} ` : ''}${t(
    'post:button_comment',
  )}`;

  const _onPressComment = () => {
    if (onPressComment) {
      onPressComment?.(postId);
    } else {
      rootNavigation.navigate(homeStack.postDetail, {
        post_id: postId,
        focus_comment: true,
      });

      // tracking event
      const eventContentReadProperties: TrackingEventContentReadProperties = {
        content_type: PostType.POST,
        action: TrackingEventContentReadAction.COMMENT,
      };
      trackEvent({ event: TrackingEvent.CONTENT_READ, properties: eventContentReadProperties });
    }
  };

  if (isLite) {
    return (
      <ContentFooterLite
        {...props}
        id={postId}
        reactionsCount={getTotalReactions(reactionsCount, 'user')}
        commentsCount={commentsCount}
        totalUsersSeen={totalUsersSeen}
      />
    );
  }

  return (
    <ContentFooter
      {...props}
      canReact={canReact}
      reactionsCount={reactionsCount}
      labelButtonComment={labelButtonComment}
      canComment={canComment}
      onPressComment={_onPressComment}
    />
  );
};

export default PostFooter;
