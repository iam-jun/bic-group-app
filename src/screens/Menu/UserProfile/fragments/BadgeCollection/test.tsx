import React from 'react';
import {
  View, Dimensions, FlatList, StyleSheet,
} from 'react-native';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import Avatar from '~/baseComponents/Avatar';
import Icon from '~/baseComponents/Icon';
import spacing, { borderRadius } from '~/theme/spacing';

const Grid = ({ data }: {data: any[]}) => {
  const deviceWidth = Dimensions.get('window').width;

  const itemWidth = 48; // Constant item width
  const numColumns = Math.floor((deviceWidth - 10 * 2) / (itemWidth + 6 * 2));
  const rowWidth = numColumns * (itemWidth + 6 * 2);
  const paddingLeft = (deviceWidth - (10 * 2) - rowWidth) / 2;

  const theme: ExtendedTheme = useTheme();
  const { colors } = theme;
  const styles = themeStyles(theme);

  const isSelected = false;

  const renderGridItem = ({ item, index }: any) =>
    // Render each grid item here
    (
      <View
        key={index}
        style={{
          margin: 6,
          borderWidth: isSelected ? 2 : 1,
          borderColor: isSelected ? colors.purple50 : colors.neutral5,
          borderRadius: borderRadius.pill,
        }}
      >
        <Avatar.Medium
          isRounded
          source={{ uri: item?.iconUrl }}
        />
        {isSelected
          ? (
            <View style={styles.iconChose}>
              <Icon size={10} icon="Check" tintColor={colors.white} />
            </View>
          )
          : null}
      </View>
    );

  return (
    <View style={{
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      paddingHorizontal: 10,
      paddingBottom: 10,
    }}
    >

      <View style={{
        // paddingLeft,
      }}
      >
        <FlatList
          data={data}
          renderItem={renderGridItem}
          keyExtractor={(item, index) => index.toString()}
          numColumns={numColumns}
        />
      </View>
    </View>
  );
};

const themeStyles = (theme: ExtendedTheme) => {
  const { colors } = theme;

  return StyleSheet.create({
    container: {
      flex: 1,
      paddingTop: spacing.padding.large,
    },
    iconChose: {
      backgroundColor: colors.purple50,
      borderRadius: borderRadius.pill,
      height: 20,
      width: 20,
      alignItems: 'center',
      justifyContent: 'center',
      position: 'absolute',
      zIndex: 1,
      right: 0,
    },
  });
};

export default Grid;
