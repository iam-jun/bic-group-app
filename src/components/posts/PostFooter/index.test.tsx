import React from 'react';
import { cleanup, renderWithRedux, store } from '~/test/testUtils';
import PostFooter from '.';

afterEach(() => {
  cleanup();
  store.clearActions();
});

describe('PostFooter component', () => {
  it('should render correctly', () => {
    const rendered = renderWithRedux(
      <PostFooter postId="test-id" labelButtonComment="Comment" reactionsCount={[]} />,
    );
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
