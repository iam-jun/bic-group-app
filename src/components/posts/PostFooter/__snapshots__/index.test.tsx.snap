// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostFooter component should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
      Object {
        "height": 28,
        "marginBottom": 8,
      },
    ]
  }
  testID="content_footer.cannot_reaction_view"
>
  <Text
    style={
      Array [
        Object {
          "color": "#212745",
          "fontFamily": "BeVietnamPro-Light",
          "fontSize": 13,
        },
        Object {
          "color": "#BAC0DE",
        },
        undefined,
      ]
    }
  >
    React and comment of this post are not available
  </Text>
</View>
`;
